k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -139,71 +121,108 @@ private[k8s] abstract class Watchable[O <: KubeObject: TypeReference, W <: Watch
+
             })
-            val watchStream = Json.readStream[W](rsp.reader, Api.BufSize)
-              // it would be nice to not have to flat map over the watch
-              // stream, but this is necessary to handle the K8s bug where
-              // erorrs have the incorrect status code.
-              .flatMap {
-                // special case to handle Kubernetes bug where "too old
-                // resource version" errors are returned with status code 200
-                // rather than status code 410.
-                // see https://github.com/kubernetes/kubernetes/issues/35068
-                // for details.
-                case e: Watch.Error[O] if e.status.code.contains(410) =>
-                  log.debug(
-                    "k8s returned 'too old resource version' error with " +
-                      "incorrect HTTP status code, restarting watch"
-                  )
-                  _resourceVersionTooOld()
-                case event => AsyncStream.fromOption(Some(event))
-              }
-            watchStream ++ // if the stream ends (k8s will kill connections after ~30m), restart it)
-              AsyncStream.fromFuture {
-                // It's safe to call lastOption here, because we're after the `++` operator, so
-                // this code will not be executed until `watchStream` has completed.
-                val currentVersion = watchStream.lastOption.map(_.flatMap(_.resourceVersion))
-                currentVersion.flatMap {
-                  case Some(version) => Future.value(_watch(Some(version)))
-                  case None =>
-                    // In this case, we want to try loading the initial information instead before watching again.
-                    restartWatches(labelSelector, fieldSelector).map {
-                      case (ws, ver) => AsyncStream.fromSeq(ws) ++ _watch(ver)
-                    }
-                }
-              }
-              .flatten
+
+            _processEventStream(
+              () => Json.readStream[W](rsp.reader, Api.BufSize),
+              resourceVersion
+            )
 
           case http.Status.Gone =>
             _resourceVersionTooOld()
 
           case status =>
             close.set(Closable.nop)
             log.debug("k8s failed to watch resource %s: %d %s", path, status.code, status.reason)
-            val f = Future.exception(Api.UnexpectedResponse(rsp))
-            AsyncStream.fromFuture(f)
+            val exc = Api.UnexpectedResponse(rsp)
+            state(Activity.Failed(exc))
+            Future.exception(exc)
         }
       }
-        // Ignore any cases where we receive a resource version lower
-        // than the last received resource version.
-        .increasingOnly
+    }
+
+    def _processEventStream(
+      stream: () => AsyncStream[W],
+      resourceVersion: Option[String],
+      previousEvent: Option[W] = None
+    ): Future[Unit] = {
+      stream().uncons.flatMap {
+        case Some((w, ws)) =>
```

Comment:
Take it or leave it, but I think this _might_ be a little clearer if we collapsed the nested `match` expression into one `match` with guards, like:
```scala
// special case to handle Kubernetes bug where "too old
// resource version" errors are returned with status code 200
// rather than status code 410.
// see https://github.com/kubernetes/kubernetes/issues/35068
// for details
case Some((e: Watch.Error[0], ws))  if e.status.code.contains(410) =>
  log.debug(
   "k8s returned 'too old resource version' error with " +
     "incorrect HTTP status code, restarting watch"
  )
  _resourceVersionTooOld()
  
case Some((event, ws)) =>
  import Ordering.Implicits._
  if (previousEvent.forall((_: W) < event)) {
    state(Activity.Ok(event))
    _processEventStream(ws, event.resourceVersion, Some(event))
  } else {
    // Ignore any events where we receive a resource version lower
    // than the last received resource version.
    _processEventStream(ws, resourceVersion, previousEvent)
  }

// if the stream ends (k8s will kill connections after ~30m), restart it.
case None if resourceVersion.isDefined => _watch(resourceVersion)

// In this case, we want to try loading the initial information instead before
// watching again.
case None => _resourceVersionTooOld()
```

Reply:
(Also rephrased the comments slightly for clarity)